### PR TITLE
Fix BN_mod_exp_consttime so it does not produce unreduced result

### PR DESCRIPTION
**Pull request by t8m**
The BN_mod_exp_consttime can produce `result == modulus` instead of 0. This was reported by @guidovranken.

Original fix was developed by @davidben for BoringSSL.

### CODE REVIEW

File: crypto/bn/rsaz_exp.h [hunk 1]
```diff
@@ -52,6 +54,29 @@ int ossl_rsaz_mod_exp_avx512_x2(BN_ULONG *res1,
...
+        r[i] = constant_time_select_64(mask, a[i], b[i]);
+    }
+}
+
+static ossl_inline BN_ULONG bn_reduce_once_in_place(BN_ULONG *r,
+                                                    BN_ULONG carry,
+                                                    const BN_ULONG *m,
+                                                    void *tmp, size_t num)
+{
+    BN_ULONG *tmpbn = tmp;
```
Comment (davidben) [inline crypto/bn/rsaz_exp.h:73]: Using `void*` and then casting to `BN_ULONG*` instead the function seems likely to cause strict aliasing violations later on. Seems better to use `BN_ULONG*` so the type system will actually notice mistakes.

File: crypto/bn/rsaz_exp.c [hunk 2]
```diff
@@ -309,6 +311,8 @@ void RSAZ_512_mod_exp(BN_ULONG result[8],
     /* from Montgomery */
     rsaz_512_mul_by_one(result, temp, m, k0);
 
+    bn_reduce_once_in_place(result, /*carry=*/0, m, storage, 8);
```
Comment (davidben) [inline crypto/bn/rsaz_exp.c:314]: BoringSSL doesn't have this code---we don't consider RSA-1024 worth optimizing for these days---so I don't know off-hand whether it is correct. But supposing it's the same algorithm as rsaz_1024 (RSA-2048), the fix looks right.

File: crypto/bn/rsaz_exp_x2.c [hunk 3]
```diff
@@ -257,6 +257,9 @@ int ossl_rsaz_mod_exp_avx512_x2(BN_ULONG *res1,
     from_words52(res1, factor_size, rr1_red);
     from_words52(res2, factor_size, rr2_red);
 
+    bn_reduce_once_in_place(res1, /*carry=*/0, m1, storage, factor_size);
+    bn_reduce_once_in_place(res2, /*carry=*/0, m2, storage, factor_size);
```
[the remark this answers is not in the document]
Comment (t8m) [reply crypto/bn/rsaz_exp_x2.c:261]: Yeah, someone who fully understands the math should verify if this is needed. @amatyuko-intc perhaps?

File: crypto/bn/rsaz_exp.h [hunk 4]
```diff
@@ -52,6 +54,29 @@ int ossl_rsaz_mod_exp_avx512_x2(BN_ULONG *res1,
...
+{
+    size_t i;
+
+    for (i = 0; i < num; i++) {
+        r[i] = constant_time_select_64(mask, a[i], b[i]);
+    }
+}
+
+static ossl_inline BN_ULONG bn_reduce_once_in_place(BN_ULONG *r,
+                                                    BN_ULONG carry,
```
[the remark this answers is not in the document]
Comment (t8m) [reply crypto/bn/rsaz_exp.h:69]: Well, not any time soon, but at some point we might want to if we decide to fix the BN const time handling.

File: crypto/bn/rsaz_exp.h (hunk 1 above)
Comment (t8m) [reply crypto/bn/rsaz_exp.h:73]: As this is just a temporary storage area for this function and we do not care about the contents after the return this should not be a problem?

**Comment by t8m** [comment]
Thanks @davidben for the review

File: crypto/bn/rsaz_exp.h (hunk 1 above)
Comment (davidben) [reply crypto/bn/rsaz_exp.h:73]: Even as temporary storage, It is UB in C to access an object with the wrong pointer type. Some violations may be more or less likely to be mangled by the compiler, but I think it's safest to not tempt fate.

Comment (t8m) [reply crypto/bn/rsaz_exp.h]: OK

**Review by mattcaswell — APPROVED** [review]

**Comment by t8m** [comment]
Merged to master and 3.0 branches. Thank you for the reviews.

File: crypto/bn/rsaz_exp_x2.c (hunk 3 above)
[the remark this answers is not in the document]
Comment (xry111) [reply crypto/bn/rsaz_exp_x2.c:261]: These two lines is causing #18625.

**Pull request closed without merging**